kernels/rviv/3/bli_rviv_utils.h

```diff
@@ -34,4 +34,32 @@
 
 #include "blis.h"
 
-uint32_t num_fp32_per_vector();
+inline uint32_t num_fp32_per_vector(void)
+{
+    uint32_t velem = 0;
+    __asm__ volatile
+    (
+    " li a5, 256                         \n\t"
+    " vsetvli a4, a5, e32, m1, ta, ma    \n\t"
+    " mv %[velem], a4                    \n\t"
+    : [velem] "=r" (velem)
+    :
+    : "a4", "a5"
+    );
```

Comment:
```suggestion
inline uint32_t get_vlenb(void)
{
    uint32_t vlenb = 0;
    asm volatile (
       " csrr %0, vlenb    \n\t"    // vector length in bytes
      : "=r" (vlenb)
    ); 
    return vlenb;
}
inline uint32_t num_fp32_per_vector(void)
{
    return get_vlenb() / sizeof(float);
}
inline uint32_t num_fp64_per_vector(void)
{
   return get_vlenb() / sizeof(double);
}
inline uint32_t num_scmplx_per_vector(void)
{
   return get_vlenb() / sizeof(scomplex);
}
inline uint32_t num_dcmplx_per_vector(void)
{
   return get_vlenb() / sizeof(dcomplex);
}
```
There is one drawback with this solution: csrr is an atomic read, so likely not that fast

Reply:
Do we really need four functions for `float`, `double`, `scomplex` and `dcomplex` ?  VLEN is guaranteed to be a power of 2 so if we divide by 4, 8, or 16 and then multiply back the size, we will have an identity operation and can use VLEN directly instead of dividing it and multiplying it back.  By moving all of the multiplication / shifts out of the assembly code into the C, the multiplications are much simpler and the division cancels out.

Why is this better than `vsetvli vlen, zero, ...`?  By atomic do you mean A extension is required or simply that `csrr` is atomic?

Reply:
> Do we really need four functions for float, double, scomplex and dcomplex

I understood your comment that you wanted to have versions for scomplex and dcomplex. If you prefer to multiply or divide the output to adjust for the data type, sure. That is what is done in the context file.

csrr appears to be the only instruction by which vlenb can be read off. My understanding is csrr is not part of A, but simply an atomic instruction that is designed to read machine data. For vlenb I do not see why it is necessary to have this as an atomic read. I assume that this is inherited behavior from reading out other machine data where atomic read and write behavior is needed.

Reply:
If you see my latest changes, I use `get_vlenb()` directly to compute byte strides. I only divide it in the context routine, where it already explains why `float` is used.